File: LeetCode/567-permutation-in-string/567-permutation-in-string.py
from typing import List


class Solution:
    # # slow
    # def checkInclusion(self, s1: str, s2: str) -> bool:
    #     l1 = len(s1)
    #     l2 = len(s2)

    #     if l2 < l1:
    #         return False

    #     char_dict_s1 = {}
    #     for e in s1:
    #         char_dict_s1[e] = char_dict_s1.get(e, 0) + 1

    #     for idx in range(l2 - l1 + 1):
    #         char_dict_s2 = {}
    #         for e in s2[idx:idx+l1]:
    #             char_dict_s2[e] = char_dict_s2.get(e, 0) + 1

    #         if char_dict_s2 == char_dict_s1:
    #             return True

    #     return False

    # # fast
    # def checkInclusion(self, s1: str, s2: str) -> bool:
    #     n = len(s1)
    #     m = len(s2)
    #     start = 0
    #     if n > m:
    #         return False

    #     s1_counts = [0] * 26
    #     s2_counts = [0] * 26
    #     for i in range(n):
    #         s1_counts[ord(s1[i]) - ord('a')] += 1
    #         s2_counts[ord(s2[i]) - ord('a')] += 1

    #     if s1_counts == s2_counts:
    #         return True
    #     for j in range(n, m):
    #         s2_counts[ord(s2[j]) - ord('a')] += 1
    #         s2_counts[ord(s2[start]) - ord('a')] -= 1
    #         start += 1

    #         if s1_counts == s2_counts:
    #             return True

    #     return False

    def checkInclusion(self, s1: str, s2: str) -> bool:
        l1 = len(s1)
        l2 = len(s2)

        if l2 < l1:
            return False

        s1_dict = {chr(97+i): 0 for i in range(26)}
        s2_dict = {chr(97+i): 0 for i in range(26)}

        for i in range(l1):
            s1_dict[s1[i]] = s1_dict.get(s1[i], 0) + 1
            s2_dict[s2[i]] = s2_dict.get(s2[i], 0) + 1

        if s1_dict == s2_dict:
            return True

        start = 0
        for i in range(l1, l2):
            s2_dict[s2[i]] = s2_dict.get(s2[i], 0) + 1
            s2_dict[s2[start]] -= 1
            start += 1
            if s1_dict == s2_dict:
                return True

        return False


if __name__ == '__main__':
    sol = Solution()
    inputs1 = ["ab", "ab", "abc"]
    inputs2 = ["eidbaooo", "eidboaoo", "ccccbbbbaaaa"]
    outputs = [True, False, False]

    for idx in range(len(outputs)):
        ans = sol.checkInclusion(inputs1[idx], inputs2[idx])
        if ans != outputs[idx]:
            print(f'{inputs1[idx]}, {inputs2[idx]}, {ans}, {outputs[idx]}')

File: LeetCode/567-permutation-in-string/567-permutation-in-string.rs
struct Solution {}

impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let l1: usize = s1.len();
        let l2: usize = s2.len();
        if l1 > l2 {
            return false;
        }

        let mut s1_vec: Vec<i32> = vec![];
        let mut s2_vec: Vec<i32> = vec![];

        for _i in 0..26 {
            s1_vec.push(0);
            s2_vec.push(0);
            // println!("{:?}, {:?}, {:?}", _i, s1_vec.get(_i), s2_vec.get(_i));
        }

        let mut n: usize = 0;
        while n < l1 {
            s1_vec[s1.as_bytes()[n] as usize - 'a' as usize] += 1;
            s2_vec[s2.as_bytes()[n] as usize - 'a' as usize] += 1;
            n += 1;
        }

        if s1_vec == s2_vec {
            return true;
        }
        // println!("s1_vec: {:?}, s2_vec: {:?}", s1_vec, s2_vec);

        let mut pos: usize = l1;
        let mut start: usize = 0;
        while pos < l2 {
            s2_vec[s2.as_bytes()[pos] as usize - 'a' as usize] += 1;
            s2_vec[s2.as_bytes()[start] as usize - 'a' as usize] -= 1;
            if s1_vec == s2_vec {
                return true;
            }

            pos += 1;
            start += 1;
        }
        return false;
    }
}

fn main() {
    let inputs1: Vec<&str> = vec!["ab", "ab", "abc", "a"];
    let inputs2: Vec<&str> = vec!["eidbaooo", "eidboaoo", "ccccbbbbaaaa", "ab"];
    let outputs: Vec<bool> = vec![true, false, false, true];

    for (pos, _) in outputs.iter().enumerate() {
        let ans = Solution::check_inclusion(inputs1[pos].to_string(), inputs2[pos].to_string());

        if ans != outputs[pos] {
            println!(
                "{:?}, {:?}, {:?}, {:?}",
                inputs1[pos], inputs2[pos], outputs[pos], ans
            );
        }
    }
}
